Revised version of cTLS (#29)
tlswg/draft-ietf-tls-ctls: draft-ietf-tls-ctls.md

Values of type uint8, opaque values, and one-byte enums are not
affected.  We do not show the structures which only change in this way.

## Template-based Specialization

The transmission overhead in TLS 1.3 is largely contributed to by two factors, 

> namely 

Reviewer: ```suggestion
:
```

> - by the negotiation of algorithm parameters, and extensions,  as well as 
> - by the exchange of certificates. 

Reviewer: ```suggestion
- the negotiation of algorithm parameters, and extensions,  as well as 
- the exchange of certificates. 
```

+---------------+---------------+---------------+
~~~~~

By assuming that out-of-band agreements took place already prior to the start of
the cTLS protocol exchange, the amount of data exchanged can be radically reduced. 
Because different clients may use different compression templates and because multiple 
compression templates may be available for use in different deployment environments, 
a client needs to inform the server about the profile it is planning to use. The 
profile field in the ClientHello serves this purpose. 

Although the template-based specialization mechanisms described here are general,
we also include specific mechanism for certificate-based exchanges because those are

> where the most complexity and size reduction can be obtained. Most exchanges in 
> TLS 1.3 are highly optimized and do not require compression to be used. 

Reviewer: ```suggestion
where the most complexity and size reduction can be obtained. Most of the other exchanges in 
TLS 1.3 are highly optimized and do not require compression to be used. 
```

+---------------+---------------+---------------+
~~~~~

By assuming that out-of-band agreements took place already prior to the start of
the cTLS protocol exchange, the amount of data exchanged can be radically reduced. 
Because different clients may use different compression templates and because multiple 
compression templates may be available for use in different deployment environments, 
a client needs to inform the server about the profile it is planning to use. The 
profile field in the ClientHello serves this purpose. 

> While it is possible to apply the template-based serialization to the entire 
> TLS 1.3 protocol, this version of the specification focuses on the certificate-based
> exchange. For other exchanges, the standard TLS 1.3 protocol is used instead. Most
> exchanges in TLS 1.3 are highly optimized and do not require compression to be used. 

Reviewer: ```suggestion
Although the template-based specialization mechanisms described here are general,
we also include specific mechanism for certificate-based exchanges because those are
where the most complexity and size reduction can be obtained. Most
exchanges in TLS 1.3 are highly optimized and do not require compression to be used. 
```

(e.g., ed25519). This implies a literal "signature_algorithms" extension
consisting solely of this group.

random (integer):
: indicates that the ClientHello.Random and ServerHello.Random values
are truncated to the given length. When the transcript is
reconstructed, the Random is padded to the right with 0s and the
anti-downgrade mechanism in {{RFC8446)}, Section 4.1.3 is disabled.
IMPORTANT: Using short Random values can lead to potential
attacks. The Random length MUST be less than or equal to 32 bytes.

mutualAuth (boolean):
: indicates whether the client has to be prepared to transmit a 
Certificate and a CertificateVerify message and whether the server 
has to transmit a CertificateRequest message. 

extension_order: 
: indicates in what order extensions appear in respective messages. 
This allows to omit sending the type. If there is only a single 
extension to be transmitted, then the extension length field can also

> be omitted. For example, imagine that only the KeyShare extension 

Reviewer: I think we'd have to explain how we knew that to be the case. There's no indication that this is the exclusive list.

Also, aren't all the extensions self-delimiting? If so, can we always remove the extension length?

At this point do you just want to put an OPEN ISSUE here?

(e.g., ed25519). This implies a literal "signature_algorithms" extension
consisting solely of this group.

random (integer):
: indicates that the ClientHello.Random and ServerHello.Random values
are truncated to the given values. When the transcript is
reconstructed, the Random is padded to the right with 0s and the
anti-downgrade mechanism in {{RFC8446)}, Section 4.1.3 is disabled.
IMPORTANT: Using short Random values can lead to potential
attacks. The Random length MUST be less than or equal to 32 bytes.
When this element is omitted, the random values are created by hashing 
the ephemeral public key via the hash algorithm indicated in 

> cipherSuite and truncated to 32 bytes. 

Reviewer: Don't we want the entire KeyShare extension?

Also, I don't think when it's "omitted" but rather when it's 0. That way, the default is current TLS, which seems like a good practice.

Author: I removed the part about creating the Random values via hashing of the DH public keys. We can add that again later

mutualAuth (boolean):
: indicates whether the client has to be prepared to transmit a 
Certificate and a CertificateVerify message and whether the server 

> has to transmit a CertificateRequest message. 

Reviewer: Isn't this actually "we are going to pretend CR was sent but it wasnt really"?

Author: Yes. Should we phrase it differently?

Reviewer: I think we would need to lay some groundwork earlier. Do we have a general statement for when this kind of optimization would be possible?

Author: The entire issue here is whether we want to pre-define that the client has to authenticate itself to the server. If so, then with TLS 1.3 the server has to send the CertificateRequest message. If we want to obmit this message then this configuration option was supposed to indicate this functionality. 

As a consequence, the Client would send a Certificate + CertificateVerify message without receiving a CertificateRequest message first. 

Is the term "mutualAuth" better replaced by something like "skipCertificateRequest" or "implicitCertificateRequest"? Maybe.

extension_order: 
: indicates in what order extensions appear in respective messages. 
This allows to omit sending the type. If there is only a single 
extension to be transmitted, then the extension length field can also
be omitted. For example, imagine that only the KeyShare extension 
needs to be sent in the ClientHello as the only extension. Then, 
the following structure

~~~
   28                    // Extensions.length
   33 26                 // KeyShare
     0024                // client_shares.length
       001d              // KeyShareEntry.group
       0020 a690...af948 // KeyShareEntry.key_exchange
~~~       

is compressed down to

~~~
   0020 a690...af948 // KeyShareEntry.key_exchange

> ~~~

Reviewer: This doesn't seem right unless you are pinning down the entire group.

ISTM that one could have:

- Extension order: then you just get to omit the type and length (I believe all current extensions are internally self-delimited)
- One group key shares.

And you could use one or both.

Author: I added a clarifying note.

of identity misbinding. This is, however, entirely separable
from the rest of the specification.

The idea of hashing the ephemeral public keys and their use as random 
values was introduced by Karthik Bhargavan and requires analysis. 

> 

Reviewer: This got removed, so you should remove it, right?

Author: I created a note in form of an open issue.
